transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,874 @@
+        fingerprint: fingerprint.to_owned(),
+        // NOTE: ufrag is equal to pwd.
+        ufrag: ufrag.to_owned(),
+        pwd: ufrag.to_owned(),
+    };
+    tt.render("description", &context).unwrap()
+}
+
+/// Tries to turn a WebRTC multiaddress into a [`SocketAddr`]. Returns None if the format of the
+/// multiaddr is wrong.
+pub(crate) fn multiaddr_to_socketaddr(addr: &Multiaddr) -> Option<SocketAddr> {
+    let mut iter = addr.iter();
+    let proto1 = iter.next()?;
+    let proto2 = iter.next()?;
+    let proto3 = iter.next()?;
+
+    for proto in iter {
+        match proto {
+            Protocol::P2p(_) => {} // Ignore a `/p2p/...` prefix of possibly outer protocols, if present.
+            _ => return None,
+        }
+    }
+
+    match (proto1, proto2, proto3) {
+        (Protocol::Ip4(ip), Protocol::Udp(port), Protocol::XWebRTC(_)) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        (Protocol::Ip6(ip), Protocol::Udp(port), Protocol::XWebRTC(_)) => {
+            Some(SocketAddr::new(ip.into(), port))
+        }
+        _ => None,
+    }
+}
+
+/// Transforms a byte array fingerprint into a string.
+pub(crate) fn fingerprint_to_string(f: &Cow<'_, [u8; 32]>) -> String {
+    let values: Vec<String> = f.iter().map(|x| format! {"{:02x}", x}).collect();
+    values.join(":")
+}
+
+/// Returns a SHA-256 fingerprint of the first certificate.
+///
+/// # Panics
+///
+/// Panics if the config does not contain any certificates.
+pub(crate) fn fingerprint_of_first_certificate(config: &RTCConfiguration) -> String {
+    // safe to unwrap here because we require a certificate during construction.
+    let fingerprints = config
+        .certificates
+        .first()
+        .expect("at least one certificate")
+        .get_fingerprints()
+        .expect("get_fingerprints to succeed");
+    debug_assert_eq!("sha-256", fingerprints.first().unwrap().algorithm);
+    fingerprints.first().unwrap().value.clone()
+}
+
+/// Creates a new [`SettingEngine`] and configures it.
+pub(crate) fn build_setting_engine(
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+    addr: &SocketAddr,
+    fingerprint: &str,
+) -> SettingEngine {
+    let mut se = SettingEngine::default();
+    // Set both ICE user and password to fingerprint.
+    // It will be checked by remote side when exchanging ICE messages.
+    let f = fingerprint.to_owned().replace(':', "");
+    se.set_ice_credentials(f.clone(), f);
+    se.set_udp_network(UDPNetwork::Muxed(udp_mux.clone()));
+    // Allow detaching data channels.
+    se.detach_data_channels();
+    // Set the desired network type.
+    //
+    // NOTE: if not set, a [`webrtc_ice::agent::Agent`] might pick a wrong local candidate
+    // (e.g. IPv6 `[::1]` while dialing an IPv4 `10.11.12.13`).
+    let network_type = if addr.is_ipv4() {
+        NetworkType::Udp4
+    } else {
+        NetworkType::Udp6
+    };
+    se.set_network_types(vec![network_type]);
+    se
+}
```

Comment:
Instead of smaller utility functions, such code could live on user-defined types with the type-system guiding you through, how to compose / call them.

I've found this post helpful in how to think about this: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate.